### PR TITLE
Close cleanup

### PR DESCRIPTION
**Pull request by foolswood**
An attempt to clean up what's going on around closing (especially in the tests).
I believe this covers what Leth's PR was getting at.

I'm seeing some issues in this kind of area, but am struggling to see how it hangs together.

### CODE REVIEW

**Comment by foolswood** [comment]
Any comments on this?

**Comment by foolswood** [comment]
Is this ok? I am concerned that it is going to go stale.

**Comment by dzen** [comment]
Hello @foolswood 

Thank you for your recent  PR.
We will look them closely.

File: aioamqp/protocol.py [hunk 1]
```diff
@@ -99,12 +99,13 @@ def close(self, no_wait=False, timeout=None):
...
         encoder.write_short(0)
         frame.write_frame(encoder)
-        if not no_wait:
-            yield from self.wait_closed(timeout=timeout)
+        yield from self.wait_closed(timeout=timeout)
 
     @asyncio.coroutine
     def wait_closed(self, timeout=None):
         yield from asyncio.wait_for(self.connection_closed.wait(), timeout=timeout, loop=self._loop)
+        for channel in tuple(self.channels.values()):
```
Comment (dzen) [inline aioamqp/protocol.py]: Not sure we have to cast to a tuple here so as in line 255

Comment (foolswood) [inline aioamqp/protocol.py]: The channels can change during iteration, so I think we do (test_channel_active_flow can be used to reproduce this one and test_connection_lost for the one on line 255).

**Comment by foolswood** [comment]
Bump

**Comment by foolswood** [comment]
I'm running in production with all of the patches I've submitted, no issues seen.

**Comment by RemiCardona** [comment]
Thanks for the patch, I'll take a look as time allows.

**Comment by RemiCardona** [comment]
@foolswood I've pushed a bunch of commits to master, some of them directly inspired by the work you did in this PR. We'll probably cut a new release soon as the changes are starting to pile up and I want this code to be battle tested before moving on to other, more invasive, changes.

If you could test the current master and report back, I'd really appreciate it.

Thanks again for your contributions and your patience!

Cheers

**Comment by foolswood** [comment]
As I suspect you've guessed we have been running on a fork for some months now.

At one point we were seeing bizarre intermittent disconnection issues.
A lot of this clean up work was done attempting to get to the bottom of them.
They haven't occurred for months now.
I don't know exactly what fixed it, maybe it was something in the clean up.

I'm afraid that I need features from all 3 of the PRs to be present before I would be happy to switch back to master.
The prospect of going back to a situation where we may have random disconnects without the tools to debug them is not an opportunity I cherish.

**Comment by RemiCardona** [comment]
AIUI, the last PR from you I need to pull is the "channel cancel" bug, all the other bugs should be fixed. Let me know if anything else is missing.

Thanks again.

**Pull request closed without merging**